### DISCUSSION TITLE
Add a custom field to query results. async is not supported

### DISCUSSION THREAD

**Discussion by Marshal-S** [prisma/prisma#23713] [Q&A]
### Bug description

I've come across a scenario where saving a file for a new file library query requires a minio signature, so I use the extends result to add a field, but minio signatures are asynchronous, and compute doesn't support async here
```prisma
model File {
    id                String                 @id @default(cuid())
    filename     String
}
```
```js
const prisma = new PrismaClient().$extends({
  result: {
    file: {
      url: {
        needs: { filename: true },
        compute(file) {
          //Here I need to get the signed url result asynchronously
          // minio.signUrl，Returns a promise
          return minio.signUrl(file.file)
        },
      },
    },
  },
})
```

// myFile
```prisma
model File {
    id                String                 @id @default(cuid())
    filename     String
}
```
```js
await prisma.file.findFirst()
//result
{
 id: ...,
 filename: '123123123.png',
 url: {}
}
```
//The expectation is this
```
{
 id: ...,
 filename: '123123123.png',
 url: 'http://.........'
}
```
### How to reproduce

prepare prisma.schema with the model as above
run createTestPrismaClient() as above

### Prisma information
```prisma
model File {
    id                String                 @id @default(cuid())
    filename     String
}
```
//excute
```
npx prisma db push
```
//async sign example
```js
const signUrl = () => {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            resolve('https://github.com/moment/moment')
        }, 10)
    })
}
```
//add extends
```js
const prisma = new PrismaClient().$extends({
  result: {
    file: {
      url: {
        needs: { filename: true },
        compute(file) {
          //Here I need to get the signed url result asynchronously
          // minio.signUrl，Returns a promise
          return minio.signUrl(file.file)
        },
      },
    },
  },
})

async testFIleSign() {
 const file =  await await prisma.file.findFirst()
 console.log(file)
}
```
### Environment & setup

OS: mac os
Database: mysql
Node.js version: v21.6.2

### Prisma Version

5.9.1

**Comment by ludralph** [comment]
Hello @Marshal-S 👋 

Thank you for raising this question.

The `compute` function in Prisma Client `$extends` method does not support asynchronous operations. It is designed to be synchronous and does not support Promises or async/await syntax. See this [comment](https://github.com/prisma/prisma/issues/18300#issuecomment-1468303756) in the closed issue. 

As a workaround, you might consider fetching the `file` data first using Prisma Client, and then separately obtaining the signed URL from Minio. You can then combine the data as needed. Here's an example:
```js
async function testFileSign() {
  const file = await prisma.file.findFirst();
  file.url = await minio.signUrl(file.file);
  console.log(file);
}
```
This approach separates the database query from the signing operation. 

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by Marshal-S** [reply]
Thanks for your reply.
My file is mainly in the association table, which is inconvenient to use

**Reply by Marshal-S** [reply]
When the relationship level is relatively deep or involves multiple parameters, this way is really inconvenient, is there a way

**Reply by ludralph** [reply]
Hi @Marshal-S,

 It is not currently possible to perform asynchronous operations within compute due to performance reasons. You can see this [comment](https://github.com/prisma/prisma/issues/18300#issuecomment-1570994133)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
